ls1intum/Artemis
# Feature/export cluster hierarchy

⚠️ Depends on ls1intum/Athene#26 to be deployed as well! (DONE)

---




### Checklist

### Motivation and Context

To offer automatic feedback to text blocks which were:
- Not assigned to any cluster in the initial clustering
- Was assigned to a cluster, but it does not have an existing feedback yet

The UML class diagram and the new database tables can be seen in this issue:
[https://github.com/ls1intum/Artemis/issues/1782](url)

### Description

- Extended the response from TextSimilarityClusteringService to include the clusterTree and distanceMatrix of blocks
- Moved all text exercise related domain classes to a new "text" package
- Added corresponding entities and repositories
     - TextTreeNode
     - TextTreeNodeRepository
     - TextPairwiseDistance
     - TextPairwiseDistanceRepository
- Modified TextSimilarityClusteringServiceTest to mock the new response of Athene
- Added TextClusteringServiceTest to test the new operations on the database

### Testing
Before testing make sure that the latest version of **Athene is running** with all of its components!
There is **no new feature** added in this PR, only new database entities and repositories are added, and some existing ones are slightly modified. The new test TextClusteringServiceTest.java covers all queries in the new repositories. 

**Steps for Testing Locally:**
1. Start Athene using "docker-compose up -d" 
2. Run Artemis Server (automaticText enabled) and Client
3. Create a text exercise with automatic feedback suggestions enabled and due date a couple of minutes later
4. Submit a text submission with >= 10 sentences (a couple of paragraphs copied from wikipedia etc.)
5. Wait for the due date of the exercise to pass so that clustering can be performed.
6. Check the database tables: text_block, text_cluster, text_pairwise_distance, text_tree_node should have corresponding entries

### Test Coverage
service.TextClusteringService - 92% (Previously 44%)

domain.text.TextTreeNode - 83% (New Entity)
domain.text.TextPairwiseDistance  - 83% (New Entity)
domain.text.TextExercise - 93% (Previously 92%)
domain.text.TextBlock - 90% (Previously 89%)
domain.text.TextCluster - 73% (Previously 70%)
